# 2-Wrangling

This sections aims to deal with outliers and missing data.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split 

In [2]:
loan_data=pd.read_csv("C:\\Users\\yfawz\\OneDrive\\Desktop\\load_default_prediction\\data\\raw\\loan.csv", low_memory=False)

In [3]:
feature_desc=pd.read_excel("C:\\Users\\yfawz\\OneDrive\\Desktop\\load_default_prediction\\data\\raw\\LCDataDictionary.xlsx")

In [4]:
loan_data.isnull().sum()

id                                            2260668
member_id                                     2260668
loan_amnt                                           0
funded_amnt                                         0
funded_amnt_inv                                     0
term                                                0
int_rate                                            0
installment                                         0
grade                                               0
sub_grade                                           0
emp_title                                      166969
emp_length                                     146907
home_ownership                                      0
annual_inc                                          4
verification_status                                 0
issue_d                                             0
loan_status                                         0
pymnt_plan                                          0
url                         

In [5]:
loan_data.columns.values

array(['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'term', 'int_rate', 'installment', 'grade', 'sub_grade',
       'emp_title', 'emp_length', 'home_ownership', 'annual_inc',
       'verification_status', 'issue_d', 'loan_status', 'pymnt_plan',
       'url', 'desc', 'purpose', 'title', 'zip_code', 'addr_state', 'dti',
       'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths',
       'mths_since_last_delinq', 'mths_since_last_record', 'open_acc',
       'pub_rec', 'revol_bal', 'revol_util', 'total_acc',
       'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d',
       'last_credit_pull_d', 'collections_12_mths_ex_med',
       'mths_since_last_major_derog', 'policy_code', 'application_type',
       'annual_inc_joint', 'dti_joint', 'verification_status_joint

There are features in this dataset that are not required, so let's remove examine which ones we need and not, one by one.

But before we do that, we will split the set into train and test splits.

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(loan_data.values, loan_data.loan_status, test_size=0.25

In [3]:
#set up new dataframe
train_df=loan_data.copy()

NameError: name 'loan_data' is not defined

We will start by removing columns that have all null values.

In [7]:
null_coumns=loan_data.columns[loan_data.isnull().sum()==len(loan_data)]

In [8]:
null_coumns

Index(['id', 'member_id', 'url'], dtype='object')

We see that following columns can be dropped as they are all null values and are insignificant:

1. 'id'
2. 'member_id'
3. 'url'



In [9]:
new_df_1=new_df.drop(null_coumns, axis=1)

We see that there are a few columns with a significant number of null values. We will set a threshold of 75%. If column values are more than 75% null, they will be dropped as they will not provide much significance when predicting.

In [26]:
null_columns_75=new_df_1.columns[(new_df_1.isnull().sum()/len(new_df_1))>=0.75]

In [27]:
null_columns_75

Index(['desc', 'mths_since_last_record', 'annual_inc_joint', 'dti_joint',
       'verification_status_joint', 'mths_since_recent_bc_dlq',
       'revol_bal_joint', 'sec_app_earliest_cr_line', 'sec_app_inq_last_6mths',
       'sec_app_mort_acc', 'sec_app_open_acc', 'sec_app_revol_util',
       'sec_app_open_act_il', 'sec_app_num_rev_accts',
       'sec_app_chargeoff_within_12_mths',
       'sec_app_collections_12_mths_ex_med',
       'sec_app_mths_since_last_major_derog', 'hardship_type',
       'hardship_reason', 'hardship_status', 'deferral_term',
       'hardship_amount', 'hardship_start_date', 'hardship_end_date',
       'payment_plan_start_date', 'hardship_length', 'hardship_dpd',
       'hardship_loan_status', 'orig_projected_additional_accrued_interest',
       'hardship_payoff_balance_amount', 'hardship_last_payment_amount',
       'debt_settlement_flag_date', 'settlement_status', 'settlement_date',
       'settlement_amount', 'settlement_percentage', 'settlement_term'],
      d

Just by looking at the columns, none of them would be significant in predicting lateness.

In [50]:
new_df_2=new_df_1.drop(null_columns_75, axis=1)

In [29]:
new_df_2.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0       2500         2500           2500.0   36 months     13.56        84.92   
1      30000        30000          30000.0   60 months     18.94       777.23   
2       5000         5000           5000.0   36 months     17.97       180.69   
3       4000         4000           4000.0   36 months     18.94       146.51   
4      30000        30000          30000.0   60 months     16.14       731.78   

  grade sub_grade       emp_title emp_length  ... percent_bc_gt_75  \
0     C        C1            Chef  10+ years  ...              0.0   
1     D        D2     Postmaster   10+ years  ...              0.0   
2     D        D1  Administrative    6 years  ...              0.0   
3     D        D2   IT Supervisor  10+ years  ...            100.0   
4     C        C4        Mechanic  10+ years  ...              0.0   

   pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort  \
0                   1.0       0.0         60124.0           16901.0   
1                   1.0       0.0        372872.0           99468.0   
2                   0.0       0.0        136927.0           11749.0   
3                   0.0       0.0        385183.0           36151.0   
4                   0.0       0.0        157548.0           29674.0   

  total_bc_limit total_il_high_credit_limit hardship_flag disbursement_method  \
0        36500.0                    18124.0             N                Cash   
1        15000.0                    94072.0             N                Cash   
2        13800.0                    10000.0             N                Cash   
3         5000.0                    44984.0             N                Cash   
4         9300.0                    32332.0             N                Cash   

  debt_settlement_flag  
0                    N  
1                    N  
2                    N  
3                    N  
4                    N  

[5 rows x 105 columns]

Our features are now 105, down from 145, a 27.5% reduction.

We now need to deal with all the missing values. Let's find the columsn with missing values.

In [51]:
columns_nan=new_df_2.columns[new_df_2.isnull().sum()!=0].values

In [52]:
len(columns_nan)

73

In [53]:
columns_nan

array(['emp_title', 'emp_length', 'annual_inc', 'title', 'zip_code',
       'dti', 'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths',
       'mths_since_last_delinq', 'open_acc', 'pub_rec', 'revol_util',
       'total_acc', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d',
       'collections_12_mths_ex_med', 'mths_since_last_major_derog',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m',
       'max_bal_bc', 'all_util', 'total_rev_hi_lim', 'inq_fi',
       'total_cu_tl', 'inq_last_12m', 'acc_open_past_24mths',
       'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_inq',
       'mths_since_recent_revol_delinq', 'num_accts_e

There are 73 columns with nan values, we need to deal with them.

1. 'emp_title', Employment title. The entries are strings. If there are Nan's, they will be classified them as "Not Given"
2. 'emp_length',Employment Length. The entries are strings. If there are Nan's, they we will classified them as "Not Given"
3. 'annual_inc', Annual Income. The entries are  floats. If there are Nan's, they will be classified as 0
4. 'title', Purpose Title is just another column of purpose but human readable, so this column will be removed.
5. 'zip_code', Zip code is that of loanee addresses. The entries are strings. Nan entries will be replaced with '000xx'
6. 'dti' is debt-to-income ratio. This is a significant ratio. Entries are floats. Nan's will be replaced with the mean. 

In [54]:
new_df_2['emp_title']=new_df_2['emp_title'].fillna("Not Given")

In [55]:
new_df_2['emp_length']=new_df_2['emp_length'].fillna("Not Given")

In [56]:
new_df_2['annual_inc']=new_df_2['annual_inc'].fillna(0)

In [57]:
new_df_2.drop('title', inplace=True, axis=1)

In [61]:
new_df_2['zip_code']=new_df_2['zip_code'].fillna('000xx')

In [90]:
new_df_2['dti']=new_df_2['dti'].fillna(np.mean(new_df_2['dti']))

Let's refresh so we have a less crowded list. As the other columns will require us to see the description what they are precisely.

In [91]:
columns_nan_1=new_df_2.columns[new_df_2.isnull().sum()!=0].values

In [94]:
len(columns_nan_1)

67

In [96]:
columns_nan_1

array(['delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths',
       'mths_since_last_delinq', 'open_acc', 'pub_rec', 'revol_util',
       'total_acc', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d',
       'collections_12_mths_ex_med', 'mths_since_last_major_derog',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m',
       'max_bal_bc', 'all_util', 'total_rev_hi_lim', 'inq_fi',
       'total_cu_tl', 'inq_last_12m', 'acc_open_past_24mths',
       'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_inq',
       'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd',
       'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats', 'num

There are 67 remaining columns, let's start!!

1. 'delinq_2yrs is a counter of how many time a loanee was 30+ days late over the past 2 years. Entries are floats. Nan's will be replaced with 0.
2. 'earliest_cr_line'.This when the borrowers first credit line was opened. Entries are strings with dates(Month-Year). If it's Nan, it is most probable thay they never had a credit line. Nan's will be replaced with 'Jan-1000'.
3. 'inq_last_6mths'. This is  a count of the number of inquiries over the last 6 months. Entries are integers. Nan's will be replaced with 0.
4. 'open_acc' is the number of current open credit lines for the borrower. These are integers and are rarely zero. Nan's will be replaced with the mean.
5. 'pub_rec'is the number of deragotory public record which is a sever case(Bankruptcy, civil court etc). These values are most commonly 0. Entries will be raplced with 0.
6. 'revol_util'. This is revolving account utilization rate in %. Entries are floats. (Example (Credit Card Balance/Credit Limit)). We will replace them the mean.
7. 'total_acc' is the number of credit lines in borrowers file. This has a large range of integers(0-100). But for ease of analysis, we will use the mean.
8. 'last_pymnt_d'. This is a string of the date of the latest payment. We will replace Nan's with 'Jan-1000'.
9. 'next_pymnt_d'. This is a string with the next schedule payment date. Nan's will be replaced with 'Jan-1000'.
10. 'last_credit_pull_d'. This is the most recent month Lending Club pulled a credit check for the borrower. Nan's will be replaced with 'Jan-1000'.

In [101]:
feature_desc[feature_desc['LoanStatNew']=='delinq_2yrs']['Description'].values

array(["The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years"],
      dtype=object)

In [103]:
new_df_2['delinq_2yrs']=new_df_2['delinq_2yrs'].fillna(0)

In [104]:
feature_desc[feature_desc['LoanStatNew']=='earliest_cr_line']['Description'].values

array(["The month the borrower's earliest reported credit line was opened"],
      dtype=object)

In [108]:
new_df_2['earliest_cr_line']=new_df_2['earliest_cr_line'].fillna('Jan-1000')

In [109]:
feature_desc[feature_desc['LoanStatNew']=='inq_last_6mths']['Description'].values

array(['The number of inquiries in past 6 months (excluding auto and mortgage inquiries)'],
      dtype=object)

In [111]:
new_df_2['inq_last_6mths']=new_df_2['inq_last_6mths'].fillna(0)

In [112]:
feature_desc[feature_desc['LoanStatNew']=='open_acc']['Description'].values

array(["The number of open credit lines in the borrower's credit file."],
      dtype=object)

In [114]:
new_df_2['open_acc']=new_df_2['open_acc'].fillna(np.mean(new_df_2['open_acc']))

In [115]:
feature_desc[feature_desc['LoanStatNew']=='pub_rec']['Description'].values

array(['Number of derogatory public records'], dtype=object)

In [121]:
new_df_2['pub_rec']=new_df_2['pub_rec'].fillna(0)

In [118]:
feature_desc[feature_desc['LoanStatNew']=='revol_util']['Description'].values

array(['Revolving line utilization rate, or the amount of credit the borrower is using relative to all available revolving credit.'],
      dtype=object)

In [122]:
new_df_2['revol_util']=new_df_2['revol_util'].fillna(np.mean(new_df_2['revol_util']))

In [123]:
feature_desc[feature_desc['LoanStatNew']=='total_acc']['Description'].values

array(["The total number of credit lines currently in the borrower's credit file"],
      dtype=object)

In [125]:
new_df_2['total_acc']=new_df_2['total_acc'].fillna(np.mean(new_df_2['total_acc']))

In [126]:
feature_desc[feature_desc['LoanStatNew']=='last_pymnt_d']['Description'].values

array(['Last month payment was received'], dtype=object)

In [130]:
new_df_2['last_pymnt_d']=new_df_2['last_pymnt_d'].fillna('Jan-1000')

In [131]:
feature_desc[feature_desc['LoanStatNew']=='next_pymnt_d']['Description'].values

array(['Next scheduled payment date'], dtype=object)

In [133]:
new_df_2['next_pymnt_d']=new_df_2['next_pymnt_d'].fillna('Jan-1000')

In [135]:
feature_desc[feature_desc['LoanStatNew']=='last_credit_pull_d']['Description'].values

array(['The most recent month LC pulled credit for this loan'],
      dtype=object)

In [136]:
new_df_2['last_credit_pull_d']=new_df_2['last_credit_pull_d'].fillna('Jan-1000')

Again Let's refresh so we have a new set to work with.

In [137]:
columns_nan_2=new_df_2.columns[new_df_2.isnull().sum()!=0].values

In [139]:
len(columns_nan_2)

57

In [141]:
columns_nan_2

array(['mths_since_last_delinq', 'collections_12_mths_ex_med',
       'mths_since_last_major_derog', 'acc_now_delinq', 'tot_coll_amt',
       'tot_cur_bal', 'open_acc_6m', 'open_act_il', 'open_il_12m',
       'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il', 'il_util',
       'open_rv_12m', 'open_rv_24m', 'max_bal_bc', 'all_util',
       'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_inq',
       'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd',
       'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats', 'num_bc_tl',
       'num_il_tl', 'num_op_rev_tl', 'num_rev_accts',
       'num_rev_tl_bal_gt_0', 'num_sats', 'num_tl_120dpd_2m',
       'num_tl_30dpd', 'num_tl_90g_dpd_24m', 'num_tl_

1. 'mths_since_last_delinq'. This is the number of months since the borrower was last 30+ days late. Nan's are probably for borrowers who were never 30+ days late and will be replaced with zeros.
2. 'collections_12_mths_ex_med'. This is the number of collections made for payments that were late in the last 12 months. Nan's will be replaced with 0.
3. 'mths_since_last_major_derog'. This is the number of months since there was 90-day or worse rating. Nan's will be replaced with 0.
4. 'acc_now_delinq'. This is the number of accounts that the borrower is delinquent(30+ days late). Most of the results are 0. Nans will be replaced with 0.
5. 'tot_coll_amt'. This is the total collected amount ever owed. Entries are floats represnting dollar amounts. Nans will be replaced with zeros.
6. 'tot_cur_bal' is the total current balance owed on all accounts.Nans will be replaced with 0s.
7. 'open_acc_6m' is the number of open trades in the last 6 months.Nans will be replaced with 0s.
8. 'open_act_il' is the number of currently active installment trades. Nans will be replaced with the mean round to the nearest integer.
9. 'open_il_12m' is the number of installment accounts opened in the last 12 months.Nans will be replaced with the mean round to the nearest integer.
10. 'open_il_24m' is the number of installment accounts opened in the last 12 months.Nans will be replaced with the mean round to the nearest integer.
11. 'mths_since_rcnt_il' number of months since most recent installment account opened.Nans will be replaced with the mean round to the nearest integer.
12. 'total_bal_il'. Current balance of only installment accounts. Nans will be replaced with the mean.
13. 'il_util'. Ratio of current balance to credit limit on installment acounts.Nans will be replaced with the mean.
14. 'open_rv_12m'. Number of revoliving accounts opened in the last 12 months. Nans will be replaced with the mean round to the nearest integer..
15. 'open_rv_24m'. Number of revoliving accounts opened in the last 24 months. Nans will be replaced with the mean round to the nearest integer.
16. 'max_bal_bc'. Maximum current balanced owed on all revolving accounts.Nans will be replaced with the mean.
17. 'all_util'. This is the balance to credit limit ratio on all trades.Nans will be replaced with the mean.

In [207]:
feature_desc[feature_desc['LoanStatNew']=='mths_since_last_delinq']['Description'].values

array(["The number of months since the borrower's last delinquency."],
      dtype=object)

In [208]:
new_df_2['mths_since_last_delinq']=new_df_2['mths_since_last_delinq'].fillna(0)

In [209]:
feature_desc[feature_desc['LoanStatNew']=='collections_12_mths_ex_med']['Description'].values

array(['Number of collections in 12 months excluding medical collections'],
      dtype=object)

In [210]:
new_df_2['collections_12_mths_ex_med']=new_df_2['collections_12_mths_ex_med'].fillna(0)

In [211]:
feature_desc[feature_desc['LoanStatNew']=='mths_since_last_major_derog']['Description'].values

array(['Months since most recent 90-day or worse rating'], dtype=object)

In [212]:
new_df_2['mths_since_last_major_derog']=new_df_2['mths_since_last_major_derog'].fillna(0)

In [213]:
feature_desc[feature_desc['LoanStatNew']=='acc_now_delinq']['Description'].values

array(['The number of accounts on which the borrower is now delinquent.'],
      dtype=object)

In [214]:
new_df_2['acc_now_delinq']=new_df_2['acc_now_delinq'].fillna(0)

In [215]:
feature_desc[feature_desc['LoanStatNew']=='tot_coll_amt']['Description'].values

array(['Total collection amounts ever owed'], dtype=object)

In [216]:
new_df_2['tot_coll_amt']=new_df_2['tot_coll_amt'].fillna(0)

In [217]:
feature_desc[feature_desc['LoanStatNew']=='tot_cur_bal']['Description'].values

array(['Total current balance of all accounts'], dtype=object)

In [218]:
new_df_2['tot_cur_bal']=new_df_2['tot_cur_bal'].fillna(0)

In [219]:
feature_desc[feature_desc['LoanStatNew']=='open_acc_6m']['Description'].values

array(['Number of open trades in last 6 months'], dtype=object)

In [220]:
new_df_2['open_acc_6m']=new_df_2['open_acc_6m'].fillna(0)

In [221]:
feature_desc[feature_desc['LoanStatNew']=='open_act_il']['Description'].values

array(['Number of currently active installment trades'], dtype=object)

In [222]:
new_df_2['open_act_il']=new_df_2['open_act_il'].fillna(np.round(np.mean(new_df_2['open_act_il']),0))

In [223]:
np.round(np.mean(new_df_2['open_act_il']),0)

3.0

In [224]:
feature_desc[feature_desc['LoanStatNew']=='open_il_12m']['Description'].values

array(['Number of installment accounts opened in past 12 months'],
      dtype=object)

In [225]:
new_df_2['open_il_12m']=new_df_2['open_il_12m'].fillna(np.round(np.mean(new_df_2['open_il_12m']),0))

In [226]:
feature_desc[feature_desc['LoanStatNew']=='open_il_24m']['Description'].values

array(['Number of installment accounts opened in past 24 months'],
      dtype=object)

In [227]:
new_df_2['open_il_24m']=new_df_2['open_il_24m'].fillna(np.round(np.mean(new_df_2['open_il_24m']),0))

In [228]:
feature_desc[feature_desc['LoanStatNew']=='mths_since_rcnt_il']['Description'].values

array(['Months since most recent installment accounts opened'],
      dtype=object)

In [229]:
new_df_2['mths_since_rcnt_il']=new_df_2['mths_since_rcnt_il'].fillna(np.round(np.mean(new_df_2['mths_since_rcnt_il']),0))

In [230]:
feature_desc[feature_desc['LoanStatNew']=='total_bal_il']['Description'].values

array(['Total current balance of all installment accounts'], dtype=object)

In [231]:
new_df_2['total_bal_il']=new_df_2['total_bal_il'].fillna(np.mean(new_df_2['total_bal_il']))

In [232]:
feature_desc[feature_desc['LoanStatNew']=='il_util']['Description'].values

array(['Ratio of total current balance to high credit/credit limit on all install acct'],
      dtype=object)

In [233]:
new_df_2['il_util']=new_df_2['il_util'].fillna(np.mean(new_df_2['il_util']))

In [234]:
feature_desc[feature_desc['LoanStatNew']=='open_rv_12m']['Description'].values

array(['Number of revolving trades opened in past 12 months'],
      dtype=object)

In [235]:
new_df_2['open_rv_12m']=new_df_2['open_rv_12m'].fillna(np.round(np.mean(new_df_2['open_rv_12m']),0))
new_df_2['open_rv_24m']=new_df_2['open_rv_24m'].fillna(np.round(np.mean(new_df_2['open_rv_24m']),0))

In [236]:
feature_desc[feature_desc['LoanStatNew']=='max_bal_bc']['Description'].values

array(['Maximum current balance owed on all revolving accounts'],
      dtype=object)

In [237]:
new_df_2['max_bal_bc']=new_df_2['max_bal_bc'].fillna(np.mean(new_df_2['max_bal_bc']))

In [238]:
feature_desc[feature_desc['LoanStatNew']=='all_util']['Description'].values

array(['Balance to credit limit on all trades'], dtype=object)

In [239]:
new_df_2['all_util']=new_df_2['all_util'].fillna(np.mean(new_df_2['all_util']))

Once again, let's refresh to clear the full columns.

In [240]:
columns_nan_3=new_df_2.columns[new_df_2.isnull().sum()!=0].values

In [241]:
len(columns_nan_3)

40

In [242]:
columns_nan_3

array(['total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_inq',
       'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd',
       'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats', 'num_bc_tl',
       'num_il_tl', 'num_op_rev_tl', 'num_rev_accts',
       'num_rev_tl_bal_gt_0', 'num_sats', 'num_tl_120dpd_2m',
       'num_tl_30dpd', 'num_tl_90g_dpd_24m', 'num_tl_op_past_12m',
       'pct_tl_nvr_dlq', 'percent_bc_gt_75', 'pub_rec_bankruptcies',
       'tax_liens', 'tot_hi_cred_lim', 'total_bal_ex_mort',
       'total_bc_limit', 'total_il_high_credit_limit'], dtype=object)

1. 'total_rev_hi_lim'. We do not have a desciption but we can assume this is the high limit of all revolinvg accounts. Nans can be replaced with means.
2. 'inq_fi'. Number of personal finance credit inquiries. Nans can be raplaced with the mean rounded to the nearest integer
3. 'total_cu_tl'. Number of finance trades in total.Nans can be raplaced with the mean rounded to the nearest integer.
4. 'inq_last_12m'. This is the number of credit inquireis in the last 12 months. Nans can be raplaced with the mean rounded to the nearest integer.
5. 'acc_open_past_24mths'. Number of trades opened in the last 24 months.Nans can be raplaced with the mean rounded to the nearest integer.
6. 'avg_cur_bal'. This is the average current balance of all accounts. Nans can be replaced with the mean.
7. 'bc_open_to_buy'. This is the total dollar value to buy on revolving bankcards. Nans can be replaced with the mean 
8. 'bc_util'. This is the ratio of total current balance to high credit/credit limit for all bankcard accounts. This can be replaced with the mean.
9. 'chargeoff_within_12_mths' Number of charges off in the last 12 months. Nans can be replced with 0.
10. 'delinq_amnt'. The past-du amount for the accounts that the borrower is now delinquent. Nans can be replaced with 0s.
11. 'mo_sin_old_il_acct'. This the number of months since oldest bank installment account opened.Nans can be raplaced with the mean rounded to the nearest integer.
12. 'mo_sin_old_rev_tl_op'. This is the number of months since oldest revolving account opened. Nans can be raplaced with the mean rounded to the nearest integer.
13. 'mo_sin_rcnt_rev_tl_op'. This is the number of months since the most recent revolving account opened.Nans can be raplaced with the mean rounded to the nearest integer.
14. 'mo_sin_rcnt_tl'. This is the number of months since the most recent account opened. Nans can be raplaced with the mean rounded to the nearest integer.  
15. 'mort_acc'. This is the number of mortgage accounts. Nans can be raplaced with the mean rounded to the nearest integer.

In [243]:
feature_desc[feature_desc['LoanStatNew']=='total_rev_hi_lim']['Description'].values

array([], dtype=object)

In [297]:
new_df_2['total_rev_hi_lim']=new_df_2['total_rev_hi_lim'].fillna(np.mean(new_df_2['total_rev_hi_lim']))

In [298]:
feature_desc[feature_desc['LoanStatNew']=='inq_fi']['Description'].values

array(['Number of personal finance inquiries'], dtype=object)

In [299]:
new_df_2['inq_fi']=new_df_2['inq_fi'].fillna(np.round(np.mean(new_df_2['inq_fi']),0))

In [300]:
feature_desc[feature_desc['LoanStatNew']=='total_cu_tl']['Description'].values

array(['Number of finance trades'], dtype=object)

In [301]:
new_df_2['total_cu_tl']=new_df_2['total_cu_tl'].fillna(np.round(np.mean(new_df_2['total_cu_tl']),0))

In [302]:
feature_desc[feature_desc['LoanStatNew']=='inq_last_12m']['Description'].values

array(['Number of credit inquiries in past 12 months'], dtype=object)

In [303]:
new_df_2['inq_last_12m']=new_df_2['inq_last_12m'].fillna(np.round(np.mean(new_df_2['inq_last_12m']),0))

In [304]:
feature_desc[feature_desc['LoanStatNew']=='acc_open_past_24mths']['Description'].values

array(['Number of trades opened in past 24 months.'], dtype=object)

In [305]:
new_df_2['acc_open_past_24mths']=new_df_2['acc_open_past_24mths'].fillna(np.round(np.mean(new_df_2['acc_open_past_24mths']),0))

In [306]:
feature_desc[feature_desc['LoanStatNew']=='avg_cur_bal']['Description'].values

array(['Average current balance of all accounts'], dtype=object)

In [307]:
new_df_2['avg_cur_bal']=new_df_2['avg_cur_bal'].fillna(np.mean(new_df_2['avg_cur_bal']))

In [308]:
feature_desc[feature_desc['LoanStatNew']=='bc_open_to_buy']['Description'].values

array(['Total open to buy on revolving bankcards.'], dtype=object)

In [309]:
new_df_2['bc_open_to_buy']=new_df_2['bc_open_to_buy'].fillna(np.mean(new_df_2['bc_open_to_buy']))

In [310]:
feature_desc[feature_desc['LoanStatNew']=='bc_util']['Description'].values

array(['Ratio of total current balance to high credit/credit limit for all bankcard accounts.'],
      dtype=object)

In [311]:
new_df_2['bc_util']=new_df_2['bc_util'].fillna(np.mean(new_df_2['bc_util']))

In [312]:
feature_desc[feature_desc['LoanStatNew']=='chargeoff_within_12_mths']['Description'].values

array(['Number of charge-offs within 12 months'], dtype=object)

In [313]:
new_df_2['chargeoff_within_12_mths']=new_df_2['chargeoff_within_12_mths'].fillna(0)

In [314]:
feature_desc[feature_desc['LoanStatNew']=='delinq_amnt']['Description'].values

array(['The past-due amount owed for the accounts on which the borrower is now delinquent.'],
      dtype=object)

In [315]:
new_df_2['delinq_amnt']=new_df_2['delinq_amnt'].fillna(0)

In [316]:
feature_desc[feature_desc['LoanStatNew']=='mo_sin_old_il_acct']['Description'].values

array(['Months since oldest bank installment account opened'],
      dtype=object)

In [317]:
new_df_2['mo_sin_old_il_acct']=new_df_2['mo_sin_old_il_acct'].fillna(np.round(np.mean(new_df_2['mo_sin_old_il_acct']),0))

In [318]:
feature_desc[feature_desc['LoanStatNew']=='mo_sin_old_rev_tl_op']['Description'].values

array(['Months since oldest revolving account opened'], dtype=object)

In [319]:
new_df_2['mo_sin_old_rev_tl_op']=new_df_2['mo_sin_old_rev_tl_op'].fillna(np.round(np.mean(new_df_2['mo_sin_old_rev_tl_op']),0))

In [320]:
feature_desc[feature_desc['LoanStatNew']=='mo_sin_rcnt_rev_tl_op']['Description'].values

array(['Months since most recent revolving account opened'], dtype=object)

In [321]:
new_df_2['mo_sin_rcnt_rev_tl_op']=new_df_2['mo_sin_rcnt_rev_tl_op'].fillna(np.round(np.mean(new_df_2['mo_sin_rcnt_rev_tl_op']),0))

In [322]:
feature_desc[feature_desc['LoanStatNew']=='mo_sin_rcnt_tl']['Description'].values

array(['Months since most recent account opened'], dtype=object)

In [323]:
new_df_2['mo_sin_rcnt_tl']=new_df_2['mo_sin_rcnt_tl'].fillna(np.round(np.mean(new_df_2['mo_sin_rcnt_tl']),0))

In [324]:
feature_desc[feature_desc['LoanStatNew']=='mort_acc']['Description'].values

array(['Number of mortgage accounts.'], dtype=object)

In [325]:
new_df_2['mort_acc']=new_df_2['mort_acc'].fillna(np.round(np.mean(new_df_2['mort_acc']),0))

AND once again....refresh

In [326]:
columns_nan_4=new_df_2.columns[new_df_2.isnull().sum()!=0].values

In [327]:
len(columns_nan_4)

25

In [328]:
columns_nan_4

array(['mths_since_recent_bc', 'mths_since_recent_inq',
       'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd',
       'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats', 'num_bc_tl',
       'num_il_tl', 'num_op_rev_tl', 'num_rev_accts',
       'num_rev_tl_bal_gt_0', 'num_sats', 'num_tl_120dpd_2m',
       'num_tl_30dpd', 'num_tl_90g_dpd_24m', 'num_tl_op_past_12m',
       'pct_tl_nvr_dlq', 'percent_bc_gt_75', 'pub_rec_bankruptcies',
       'tax_liens', 'tot_hi_cred_lim', 'total_bal_ex_mort',
       'total_bc_limit', 'total_il_high_credit_limit'], dtype=object)

1. 'mths_since_recent_bc'. Number of months since most recent bankcard account opened. Nans replaced with mean round to nearest integer.
2. 'mths_since_recent_inq' Number of months since most recent inquiry. Nans replaced with mean round to nearest integer.
3. 'mths_since_recent_revol_delinq'. Number of months since most recent revolving delinquency.Nans replaced with 0
4. 'num_accts_ever_120_pd'. Number of accounts ever 120 or more days past due. Nans replaced with 0.
5. 'num_actv_bc_tl'. Number of currently active bankcard accounts. Nans replaced with mean round to nearest integer.
6. 'num_actv_rev_tl' Number of currently active revolving trades.Nans replaced with mean round to nearest integer.
7. 'num_bc_sats'.Number of satisfactory bankcard accounts. Nans replaced with mean round to nearest integer.
8. 'num_bc_tl'.Number of bankcard accounts.  Nans replaced with mean round to nearest integer.
9. 'num_il_tl'. Number of installment accounts. Nans replaced with mean round to nearest integer.
10. 'num_op_rev_tl'. Number of open revolving accounts

In [329]:
feature_desc[feature_desc['LoanStatNew']=='mths_since_recent_bc']['Description'].values

array(['Months since most recent bankcard account opened.'], dtype=object)

In [340]:
new_df_2['mths_since_recent_bc']=new_df_2['mths_since_recent_bc'].fillna(np.round(np.mean(new_df_2['mths_since_recent_bc']),0))

In [330]:
feature_desc[feature_desc['LoanStatNew']=='mths_since_recent_inq']['Description'].values

array(['Months since most recent inquiry.'], dtype=object)

In [349]:
new_df_2['mths_since_recent_inq']=new_df_2['mths_since_recent_inq'].fillna(np.round(np.mean(new_df_2['mths_since_recent_inq']),0))

In [331]:
feature_desc[feature_desc['LoanStatNew']=='mths_since_recent_revol_delinq']['Description'].values

array(['Months since most recent revolving delinquency.'], dtype=object)

In [364]:
new_df_2['mths_since_recent_revol_delinq']=new_df_2['mths_since_recent_revol_delinq'].fillna(0)

In [332]:
feature_desc[feature_desc['LoanStatNew']=='num_accts_ever_120_pd']['Description'].values

array(['Number of accounts ever 120 or more days past due'], dtype=object)

In [347]:
new_df_2['num_accts_ever_120_pd']=new_df_2['num_accts_ever_120_pd'].fillna(0)

In [333]:
feature_desc[feature_desc['LoanStatNew']=='num_actv_bc_tl']['Description'].values

array(['Number of currently active bankcard accounts'], dtype=object)

In [350]:
new_df_2['num_actv_bc_tl']=new_df_2['num_actv_bc_tl'].fillna(np.round(np.mean(new_df_2['mths_since_recent_inq']),0))

In [334]:
feature_desc[feature_desc['LoanStatNew']=='num_actv_rev_tl']['Description'].values

array(['Number of currently active revolving trades'], dtype=object)

In [352]:
new_df_2['num_actv_rev_tl']=new_df_2['num_actv_rev_tl'].fillna(np.round(np.mean(new_df_2['num_actv_rev_tl']),0))

In [335]:
feature_desc[feature_desc['LoanStatNew']=='num_bc_sats']['Description'].values

array(['Number of satisfactory bankcard accounts'], dtype=object)

In [354]:
new_df_2['num_bc_sats']=new_df_2['num_bc_sats'].fillna(np.round(np.mean(new_df_2['num_bc_sats']),0))

In [336]:
feature_desc[feature_desc['LoanStatNew']=='num_bc_tl']['Description'].values

array(['Number of bankcard accounts'], dtype=object)

In [356]:
new_df_2['num_bc_tl']=new_df_2['num_bc_tl'].fillna(np.round(np.mean(new_df_2['num_bc_tl']),0))

In [337]:
feature_desc[feature_desc['LoanStatNew']=='num_il_tl']['Description'].values

array(['Number of installment accounts'], dtype=object)

In [358]:
new_df_2['num_il_tl']=new_df_2['num_il_tl'].fillna(np.round(np.mean(new_df_2['num_il_tl']),0))

In [338]:
feature_desc[feature_desc['LoanStatNew']=='num_op_rev_tl']['Description'].values

array(['Number of open revolving accounts'], dtype=object)

In [359]:
new_df_2['num_op_rev_tl']=new_df_2['num_op_rev_tl'].fillna(np.round(np.mean(new_df_2['num_op_rev_tl']),0))

This will be the last refresh of them all.........

In [365]:
columns_nan_5=new_df_2.columns[new_df_2.isnull().sum()!=0].values

In [366]:
len(columns_nan_5)

15

In [367]:
columns_nan_5

array(['num_rev_accts', 'num_rev_tl_bal_gt_0', 'num_sats',
       'num_tl_120dpd_2m', 'num_tl_30dpd', 'num_tl_90g_dpd_24m',
       'num_tl_op_past_12m', 'pct_tl_nvr_dlq', 'percent_bc_gt_75',
       'pub_rec_bankruptcies', 'tax_liens', 'tot_hi_cred_lim',
       'total_bal_ex_mort', 'total_bc_limit',
       'total_il_high_credit_limit'], dtype=object)

1. 'num_rev_accts'. This is the number of revolving accounts per borrower. Nans replaced with mean rounded to nearest integer.
2. 'num_rev_tl_bal_gt_0'. Number of revolving trades with balance >0. Nans replaced with mean rounded to nearest integer.
3. 'num_sats'. Number of satisfactoy accounts.  Nans replaced with mean rounded to nearest integer. 
4. 'num_tl_120dpd_2m'. Number of accounts currently 120 days past due (updated in past 2 months). Nans replaced with 0.
5. 'num_tl_30dpd'.Number of accounts currently 30 days past due (updated in past 2 months).Nans replaced with 0.
6. 'num_tl_90g_dpd_24m'. Number of accounts 90 or more days past due in last 24 months. Nans replaced with 0.
7. 'num_tl_op_past_12m'. Number of accounts opened in past 12 months.Nans replaced with mean rounded to nearest integer.
8. 'pct_tl_nvr_dlq'. Percent of trades never delinquent. Nans replaced with mean.
9. 'percent_bc_gt_75'.Percentage of all bankcard accounts > 75% of limit.Nans replaced with mean.
10. 'pub_rec_bankruptcies'.Number of public record bankruptcies. Nans replaced with 0.
11. 'tax_liens'. Number of tax liens. Nans replaced with 0.
12. 'tot_hi_cred_lim'. Total high credit/credit limit. Nans replaced with mean.
13. 'total_bal_ex_mort'. Total credit balance excluding mortgage.Nans replaced with mean.
14. 'total_bc_limit'. Total bankcard high credit/credit limit. Nans replaced with mean.
15. 'total_il_high_credit_limit'. Total installment high credit/credit limit. Nans replaced with mean.

In [368]:
feature_desc[feature_desc['LoanStatNew']=='num_rev_accts']['Description'].values

array(['Number of revolving accounts'], dtype=object)

In [371]:
new_df_2['num_rev_accts']=new_df_2['num_rev_accts'].fillna(np.round(np.mean(new_df_2['num_rev_accts']),0))

In [372]:
feature_desc[feature_desc['LoanStatNew']=='num_rev_tl_bal_gt_0']['Description'].values

array(['Number of revolving trades with balance >0'], dtype=object)

In [374]:
new_df_2['num_rev_tl_bal_gt_0']=new_df_2['num_rev_tl_bal_gt_0'].fillna(np.round(np.mean(new_df_2['num_rev_tl_bal_gt_0']),0))

In [375]:
feature_desc[feature_desc['LoanStatNew']=='num_sats']['Description'].values

array(['Number of satisfactory accounts'], dtype=object)

In [377]:
new_df_2['num_sats']=new_df_2['num_sats'].fillna(np.round(np.mean(new_df_2['num_sats']),0))

In [378]:
feature_desc[feature_desc['LoanStatNew']=='num_tl_120dpd_2m']['Description'].values

array(['Number of accounts currently 120 days past due (updated in past 2 months)'],
      dtype=object)

In [419]:
new_df_2['num_tl_120dpd_2m']=new_df_2['num_tl_120dpd_2m'].fillna(0)

In [381]:
feature_desc[feature_desc['LoanStatNew']=='num_tl_30dpd']['Description'].values

array(['Number of accounts currently 30 days past due (updated in past 2 months)'],
      dtype=object)

In [384]:
new_df_2['num_tl_30dpd']=new_df_2['num_tl_30dpd'].fillna(0)

In [385]:
feature_desc[feature_desc['LoanStatNew']=='num_tl_90g_dpd_24m']['Description'].values

array(['Number of accounts 90 or more days past due in last 24 months'],
      dtype=object)

In [387]:
new_df_2['num_tl_90g_dpd_24m']=new_df_2['num_tl_90g_dpd_24m'].fillna(0)

In [388]:
feature_desc[feature_desc['LoanStatNew']=='num_tl_op_past_12m']['Description'].values

array(['Number of accounts opened in past 12 months'], dtype=object)

In [390]:
new_df_2['num_tl_op_past_12m']=new_df_2['num_tl_op_past_12m'].fillna(np.round(np.mean(new_df_2['num_tl_op_past_12m']),0))

In [391]:
feature_desc[feature_desc['LoanStatNew']=='pct_tl_nvr_dlq']['Description'].values

array(['Percent of trades never delinquent'], dtype=object)

In [397]:
new_df_2['pct_tl_nvr_dlq']=new_df_2['pct_tl_nvr_dlq'].fillna(np.mean(new_df_2['pct_tl_nvr_dlq']))

In [398]:
feature_desc[feature_desc['LoanStatNew']=='percent_bc_gt_75']['Description'].values

array(['Percentage of all bankcard accounts > 75% of limit.'],
      dtype=object)

In [399]:
new_df_2['percent_bc_gt_75']=new_df_2['percent_bc_gt_75'].fillna(np.mean(new_df_2['percent_bc_gt_75']))

In [400]:
feature_desc[feature_desc['LoanStatNew']=='pub_rec_bankruptcies']['Description'].values

array(['Number of public record bankruptcies'], dtype=object)

In [402]:
new_df_2['pub_rec_bankruptcies']=new_df_2['pub_rec_bankruptcies'].fillna(0)

In [403]:
feature_desc[feature_desc['LoanStatNew']=='tax_liens']['Description'].values

array(['Number of tax liens'], dtype=object)

In [405]:
new_df_2['tax_liens']=new_df_2['tax_liens'].fillna(0)

In [406]:
feature_desc[feature_desc['LoanStatNew']=='tot_hi_cred_lim']['Description'].values

array(['Total high credit/credit limit'], dtype=object)

In [407]:
new_df_2['tot_hi_cred_lim']=new_df_2['tot_hi_cred_lim'].fillna(np.mean(new_df_2['tot_hi_cred_lim']))

In [408]:
feature_desc[feature_desc['LoanStatNew']=='total_bal_ex_mort']['Description'].values

array(['Total credit balance excluding mortgage'], dtype=object)

In [411]:
new_df_2['total_bal_ex_mort']=new_df_2['total_bal_ex_mort'].fillna(np.mean(new_df_2['total_bal_ex_mort']))

In [412]:
feature_desc[feature_desc['LoanStatNew']=='total_bc_limit']['Description'].values

array(['Total bankcard high credit/credit limit'], dtype=object)

In [413]:
new_df_2['total_bc_limit']=new_df_2['total_bc_limit'].fillna(np.mean(new_df_2['total_bc_limit']))

In [414]:
feature_desc[feature_desc['LoanStatNew']=='total_il_high_credit_limit']['Description'].values

array(['Total installment high credit/credit limit'], dtype=object)

In [415]:
new_df_2['total_il_high_credit_limit']=new_df_2['total_il_high_credit_limit'].fillna(np.mean(new_df_2['total_il_high_credit_limit']))

Let's do a final test to ensure that there are no Nan values in our data.

In [420]:
new_df_2.columns[new_df_2.isnull().sum()!=0]

Index([], dtype='object')

All our columns contating values and there were no evident outliers detected that were signicant enough to remove.

In [422]:
final_df=new_df_2

In [423]:
final_df.to_csv("C:\\Users\\yfawz\\OneDrive\\Desktop\\load_default_prediction\\data\\processed\\loan_wrangled.csv")